### PR TITLE
Fix #316 by labeling the cluster uid to each PVC

### LINKED ISSUE #316: [BUG] PVC is mistakenly deleted when the controller reads stale `deletionTimestamp` information
## Bug Report
We find that in a HA k8s cluster, casskop could mistakenly delete the PVC of the cassandra pod after casskop experiencing a restart. After diagnosis and inspection, we find it is caused by potential staleness in some of the apiserver.

More concretely, if casskop receives a stale update event from an apiserver saying "CassandraCluster has non-nil deletion timestamp", the controller will delete the PVCs for each DataCenter and Rack belong to current CassandraCluster by calling `DeletePVCs(ReconcileCassandraCluster, dcName, rackName)`. Inside `DeletePVCs`, PVCs to be deleted are fetched with Cluster, DataCenter, and Rack name. Note that the stale event actually comes from the deletion of a previous CassandraCluster sharing the same name (but with a different UID) as the currently running one, so the PVC of the existing CassandraCluster will be listed and deleted mistakenly.

One potential approach to fix this is to label each PVC the UID of the CassandraCluster when creating them, and list the PVC using the UID of the CassandraCluster in `ListPVC` to ensure that we always delete the right PVC.

**What did you do?**
We list the concrete steps we took that led to this issue:
1. Create a CassandraCluster `cc` with `DeletePVC` enabled. PVC will be created in the cluster.
2. Delete `cc`. Apiserver1 will send the update events with a non-nil deletion timestamp to the controller and the controller will trigger `DeletePVCs()` to delete related PVC for each rack. Meanwhile, apiserver2 is partitioned so its watch cache stops at the moment that `cc` is tagged with a deletion timestamp.
3. Create the CassandraCluster with the same name `cc` again. Now the CassandraCluster gets back with a different UID, so as its PVC. However, apiserver2 still holds the stale view that `cc` has a non-nil deletion timestamp and is about to be deleted.
4. The controller restarts after a node failure and talks to the stale apiserver2. Reading the stale update events from apiserver2 that `cc` has a deletion timestamp, the controller lists all the PVC belonging to the currently running `cc` (as mentioned above) and deletes them all.

**What did you expect to see?**
The controller should not try to delete the PVC even when reading stale `deletionTimestamp`.

**What did you see instead? Under which circumstances?**
The controllers deletes the PVC after the restart as mentioned in the above steps.

**Environment**
* casskop version:
f87c8e05c1a2896732fc5f3a174f1eb99e936907 (master branch)

* Kubernetes version information:
1.18.9

* Cassandra version:
3.11

**Possible Solution**
_We are willing to help fix this bug by issuing a PR._

As mentioned above, the bug can be avoided by labeling each PVC with the UID of the CassandraCluster and listing PVC with UID. Each CassandraCluster will always have a different UID even with the same name. So in this case the PVCs belonging to the newly created CassandraCluster will not be deleted by the stale events of the old one.

### PR DESCRIPTION
**Pull request by srteam2020**
| Q               | A
| --------------- | ---
| Bug fix?        | [X]
| Related tickets | fixes #316
| License         | Apache 2.0

### What's in this PR?
Label the cassandracluster uid to the PVC and checks the uid before deleting PVC.

### Checklist

- [ X] Implementation tested

### CODE REVIEW

File: pkg/controller/cassandracluster/testdata/cassandracluster-1DC.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 apiVersion: db.orange.com/v1alpha1
 kind: CassandraCluster
 metadata:
+  uid: cassandra-test-uid
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/testdata/cassandracluster-1DC.yaml:4]: this change is not needed

File: pkg/controller/cassandracluster/testdata/cassandracluster-2DC-dc1-rack1-sts.yaml [hunk 2]
```diff
@@ -11,6 +11,7 @@ metadata:
   labels:
     app: cassandracluster
     cassandraCluster: cassandra-demo
+    cassandracluster-uid: cassandra-test-uid
```
Comment (cscetbon) [inline pkg/controller/cassandracluster/testdata/cassandracluster-2DC-dc1-rack1-sts.yaml:14]: not needed as well

**Review by cscetbon — COMMENTED** [review]
Other than those minor comments, it looks good to me

File: pkg/controller/cassandracluster/testdata/cassandracluster-1DC.yaml (hunk 1 above)
Comment (srteam2020) [reply pkg/controller/cassandracluster/testdata/cassandracluster-1DC.yaml]: Removed

**Comment by cscetbon** [comment]
@srteam2020 i wonder if it doesn't make more sense to update the helper to create Cassandra cluster to add a uid on the object if it's not there by default. Wdyt ? That would need we don't need the uid at all in our deployment files like we will do on a real k8s cluster.
In the meantime integration tests are running

**Comment by srteam2020** [comment]
> @srteam2020 i wonder if it doesn't make more sense to update the helper to create Cassandra cluster to add a uid on the object if it's not there by default. Wdyt ? That would need we don't need the uid at all in our deployment files like we will do on a real k8s cluster.
> In the meantime integration tests are running

Yes, that definitely makes more sense to me, I'll revise that accordingly.

**Comment by cscetbon** [comment]
@srteam2020 I'll soon make a new release. It would be great if you can finish that one in order to include it. Thanks ! Lmk if you need some help

File: pkg/controller/cassandracluster/cassandra_status_test.go [hunk 3]
```diff
@@ -95,6 +96,10 @@ func HelperInitCluster(t *testing.T, name string) (*ReconcileCassandraCluster,
 	var cc api.CassandraCluster
 	yaml.Unmarshal(common.HelperLoadBytes(t, name), &cc)
 
+	if cc.ObjectMeta.UID == "" {
```
[the remark this answers is not in the document]
Comment (cscetbon) [reply pkg/controller/cassandracluster/cassandra_status_test.go:99]: last comment I swear 😬

[the remark this answers is not in the document]
Comment (srteam2020) [reply pkg/controller/cassandracluster/cassandra_status_test.go:99]: Yeah, I just don't want to overwrite that if later some developer need to specify his UID in config file.

[the remark this answers is not in the document]
Comment (srteam2020) [reply pkg/controller/cassandracluster/cassandra_status_test.go:99]: @cscetbon Do you think that can be reasonable? I can also remove the checking if you think it doesn't matter.

**Comment by srteam2020** [comment]
@cscetbon Hi, I've removed the empty UID checking accordingly, I think now it is ready to go.

**Review by cscetbon — APPROVED** [review]

**Comment by cscetbon** [comment]
Thank you @srteam2020 for your first contribution ! 👏

**Comment by cscetbon** [comment]
@srteam2020 we discovered that this PR causes issues for users upgrading from 1.1.3-release to 1.1.4-release because of that label that is missing. We need to come up with the right way to do the upgrade.

**Comment by srteam2020** [comment]
From the error msg, it seems that the label (`cassandracluster-uid`) we added was updated into some field of `Spec`, but normally, we only suppose it to be updated to `ObjectMeta.Labels`.
Looking into the [code](https://github.com/Orange-OpenSource/casskop/blob/ab9bec76d1e5ff44fe2a13208144bef27fd8740a/pkg/controller/cassandracluster/generator.go#L347) which constructs statefulset, I found the label was also set into `spec.selector` which I believe cause the problem.

I think to add some filter here to the labels to only set `cassandracluster-uid` to statefulset’s `ObjectMeta.Labels` can fix the issue. What do you think @cscetbon .

**Comment by cscetbon** [comment]
That might work. I was going to rollback the change but that could be the answer. It still means a rolling restart during an upgrade but that should be ok. Can you test such a change and check an upgrade from 1.1.3 ?

**Pull request merged**